### PR TITLE
Fix for SSH tunneler health check

### PR DESCRIPTION
**Pull request by avorima**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

If the cloud-provider is external or doesn't support installing SSH keys and the api server is configured to use the SSH tunneler, the health check will always fail. This patch will fix the health check by basically skipping to verify the last SSH key sync time when syncing SSH keys is not supported.

I know that the SSH tunneler is deprecated, but I need it until the newer proxy solution is ready. 

Fixes one mentioned problem of #59347

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Skip SSH key sync verification during health check if SSH tunneling is enabled and syncing SSH keys is not supported.
```

### CODE REVIEW

**Comment by avorima** [comment]
I'm sorry, I've been trying to reproduce the integration test error locally, but I've only been getting (seemingly) unrelated errors. I'll have to debug this further on monday.

**Comment by odinuge** [comment]
> I'm sorry, I've been trying to reproduce the integration test error locally, but I've only been getting (seemingly) unrelated errors. I'll have to debug this further on monday.

Looks like the integration test error was a flake. It passed all checks now. :smile:

**Comment by avorima** [comment]
Thanks @odinuge

**Comment by odinuge** [comment]
Ref. parent issue involvement:

/assign wojtek-t

**Comment by cheftako** [comment]
Why would CCM need a SSH Tunnel?

**Comment by avorima** [comment]
@cheftako The SSH tunneler of the kube-apiserver uses the `AddSSHKeyToAllInstances` method from the CCM to periodically resync keys. If the CCM is external or doesn't support that method, this resync is always skipped.
The problem is that the apiserver uses the last resync time for its health check. Because the time is never updated, the check eventually fails and causes a restart.

**Comment by cheftako** [comment]
@avorima  The AddSSHKeyToAllInstances method is from the cloud-provider interface (As far as I know only implemented by the Google implementation). It is not part of the CCM specific code (Cloud Controller Manager) and as far as I know should never be called by the CCM.

**Comment by avorima** [comment]
@cheftako Ah sorry, I misused the term CCM it seems. I meant the cloud-provider interface and its implementations.
You're right it's not called by the CCM code, but the SSH tunneler receives the mentioned function from a cloud-provider implementation during initialization and uses it later on.

**Comment by cheftako** [comment]
Thanks for fixing the comment. Sorry for the confusion, wanted to make sure we had not added a use case I was unaware of.

/lgtm

**Comment by avorima** [comment]
I'm not too familiar the release process, but would it be possible to cherry pick this into older versions as part of a patch? I'm mainly using 1.15, but 1.14 would be nice as well.

**Comment by cheftako** [comment]
@avorima Start by taking a look at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md. Note that you need to get approval from the appropriate branch manager for each branch you want to cherry pick into. Also only bug fixes (which this is) are supposed to get cherry picked.

**Comment by cheftako** [comment]
@avorima Do you have SSHUser defined on you Kube API Server? I suspect that you do and don't need it. If that is true then you may just be able to remove that option, restart the kube apiserver and the problem may fix itself.

**Comment by avorima** [comment]
@cheftako Thanks.

>Do you have SSHUser defined on you Kube API Server? I suspect that you do and don't need it. If that is true then you may just be able to remove that option, restart the kube apiserver and the problem may fix itself.

Yes, but AFAICT the SSHUser needs to be defined in order for the SSH tunneler to be instantiated, so I don't think I can remove that option.

**Comment by cheftako** [comment]
@avorima you are correct. I'm just not used to anyone other than Google using the SSH tunneler.

File: pkg/master/tunneler/ssh.go [hunk 1]
```diff
@@ -157,9 +157,14 @@ func (c *SSHTunneler) SecondsSinceSync() int64 {
 }
 
 func (c *SSHTunneler) SecondsSinceSSHKeySync() int64 {
-	now := c.clock.Now().Unix()
-	then := atomic.LoadInt64(&c.lastSSHKeySync)
-	return now - then
+	// If the cloud-provider doesn't support installing SSH keys this function will
+	// eventually cause the health check to fail because the time is never updated.
```
Comment (liggitt) [inline pkg/master/tunneler/ssh.go]: The ssh-user option is long-deprecated, and the fact that health checks fail on it currently mean we know successful usage is limited. I am not in favor of enabling expanded use of a deprecated option. See https://groups.google.com/d/msg/kubernetes-sig-api-machinery/bwJJ93qA99o/pRTVFaIlBAAJ

**Comment by liggitt** [comment]
/hold

is there a compelling reason to enable additional use of this feature?

**Comment by avorima** [comment]
@liggitt In my current setup, it's required to get `kubectl proxy` to work.

I'm running kube-apiservers in a different network than my worker nodes. `kubectl proxy` makes a call  to the apiservers and they, instead of connecting to the worker nodes, try to connect to the cluster they're running in.
With the SSH tunneler enabled the apiserver can access the nodes and `kubectl proxy` works.

I've seen that the [API Server Network Proxy](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190226-network-proxy.md) went alpha in 1.16, but the feature set seems too limited to be useful as of now. I'm also still using 1.15 and 1.14, so I was hoping to get this cherry picked into older versions that don't yet support the new proxy.

**Comment by liggitt** [comment]
I don't think it's a good idea to make this feature work in more environments than it currently does, but I'd be interested to hear other api-machinery maintainers' thoughts. @deads2k, @lavalamp, @cheftako?

**Comment by lavalamp** [comment]
I'm on the fence, I don't want to encourage use of this, OTOH this is an embarrassing reason for it to be broken, I'd rather have it deliberately off than on-but-broken.

@avorima have you tested & are sure that this is the only thing keeping it from working in your setup?

**Comment by cheftako** [comment]
I'm with Daniel. I don't like the idea that this feature only works for in-tree providers. 
I don't think we should extend the functionality for additional use cases but this feels more fixing a bug in the code.

**Comment by avorima** [comment]
@lavalamp Yes, I have spoken with my (dev)ops colleagues about an alternative solution. I've mentioned a few points like VPNs from the discussion that @liggitt posted. They've also tried using different CNIs.
We've concluded that currenly this is the only option.

I understand the reasoning behind deprecating this feature and will continue to watch the progress of the network proxy KEP and switch to such a setup as soon as it meets our needs.

**Comment by deads2k** [comment]
> I'm on the fence, I don't want to encourage use of this, OTOH this is an embarrassing reason for it to be broken, I'd rather have it deliberately off than on-but-broken.
> 
> @avorima have you tested & are sure that this is the only thing keeping it from working in your setup?

This feature has been deprecated for two years: https://github.com/kubernetes/kubernetes/pull/54433 .  It could be removed at this point, but we've been trying to be polite to current users by waiting until the proxy implementation is ready.  I would rather take https://github.com/kubernetes/kubernetes/pull/75330 than make it possible to allow other users to latch onto this deprecated functionality.

I recommend closing.

**Comment by deads2k** [comment]
I think the nay's have it.  I would like to see progress on the proxy implementation this release.  I don't think it's tenable to keep the ssh code in deprecation, but not deleted for much longer.

/close

**Pull request closed without merging**